lib/shrine/storage/google_cloud_storage.rb

```diff
@@ -1,129 +1,111 @@
+              ).merge(_options)
           )
         end
       end
 
+      # URL to the remote file, accepts options for customizing the URL
       def url(id, **_options)
-        # URL to the remote file, accepts options for customizing the URL
-        host = @host || "storage.googleapis.com/#{@bucket}"
-
-        "https://#{host}/#{object_name(id)}"
+        if(_options.key? :expires)
+          signed_url = presign(id, _options).url
+          if @host.nil?
+            signed_url
+          else
+            signed_url.gsub(/storage.googleapis.com\/#{@bucket}/, @host)
+          end
+        else
+          host = @host || "storage.googleapis.com/#{@bucket}"
+          "https://#{host}/#{object_name(id)}"
+        end
       end
 
+      # Downloads the file from GCS, and returns a `Tempfile`.
       def download(id)
         tempfile = Tempfile.new(["googlestorage", File.extname(id)], binmode: true)
-        storage_api.get_object(@bucket, object_name(id), download_dest: tempfile)
+        get_file(id).download tempfile.path
         tempfile.tap(&:open)
       end
 
+      # Download the remote file to an in-memory StringIO object
+      # @return [StringIO] object
       def open(id)
-        # returns the remote file as an IO-like object
-        io = storage_api.get_object(@bucket, object_name(id), download_dest: StringIO.new)
+        io = get_file(id).download
         io.rewind
         io
       end
 
+      # checks if the file exists on the storage
       def exists?(id)
-        # checks if the file exists on the storage
-        storage_api.get_object(@bucket, object_name(id)) do |_, err|
-          if err
-            if err.status_code == 404
-              false
-            else
-              raise err
-            end
-          else
-            true
-          end
-        end
+        file = get_file(id)
+        return false if file.nil?
+        file.exists?
       end
 
+      # deletes the file from the storage
       def delete(id)
-        # deletes the file from the storage
-        storage_api.delete_object(@bucket, object_name(id))
-
-      rescue Google::Apis::ClientError => e
-        # The object does not exist, Shrine expects us to be ok
-        return true if e.status_code == 404
-
-        raise e
+        file = get_file(id)
+        file.delete unless file.nil?
       end
 
+      # Deletes multiple files at once from the storage.
       def multi_delete(ids)
         batch_delete(ids.map { |i| object_name(i) })
       end
 
+      # Otherwise deletes all objects from the storage.
       def clear!
-        all_objects = storage_api.fetch_all do |token, s|
-          prefix = "#{@prefix}/" if @prefix
-          s.list_objects(
-            @bucket,
-            prefix: prefix,
-            fields: "items/name",
-            page_token: token,
-          )
-        end
-
+        prefix = "#{@prefix}/" if @prefix
+        all_objects = get_bucket.files prefix: prefix
```

Comment:
It will not work if the bucket contains a lot of objects.
See https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud/v0.44.0/google/cloud/storage/file/list
It might be a good idea to use `next` and call `batch_delete` for each page, compared to use `all` which could create a very big array.